### PR TITLE
Remove has_unitary_generator attribute from operators with non-unitar…

### PR DESCRIPTION
**Pull request by ejthomas**
…y generators

**Context:**
Operators with the `has_unitary_generator` attribute would intuitively be expected to have unitary generators, i.e. a generator $G$ satisfying $G^\dagger G = \mathbb{I}$ (up to a constant prefactor). However several operators with this attribute did not actually satisfy this.

**Description of the Change:**
Removed names of operators without unitary generators from `has_unitary_generators` constructor in `qml.ops.qubit.attributes`. A corresponding list of the operator classes (not names) with this attribute, `has_unitary_generator_types`, was removed from the same module.

The `_recursive_decomposition()` method of `qml.ops.op_math.exp.Exp` needed to be updated to loop over all operators with generators rather than just those in the `has_unitary_generator_types` list. This increases the number of possible decompositions considered. The implementation iterates over `qml.ops.qubit.__all__` to find operators and filters based on the `has_generator` property of each. This means possible decompositions are not considered in a well-defined order (`qml.ops.qubit.__all__` is instantiated from a set), so to ensure more specific decompositions are favoured over more general ones (e.g. prefer `IsingZZ` over `MultiRZ` if possible), the decompositions to check are ordered with those having `operator_cls.num_wires == AnyWires` last (following an explanatory comment about the ordering chosen for `has_unitary_generator_types`).

Unit test `test_generator_decomposition` modified in `tests/ops/op_math/test_exp.py` to remove dependency on `has_unitary_generator_types` and accommodate larger set of test cases. A new unit test `test_decompose_all_nonunitary_generator` in `tests/transforms/test_tape_expand.py` confirms that the five cases identified of operators without unitary generators being identified as unitarily generated and so not decomposed by `qml.transforms.expand_nonunitary_gen` are all now decomposed to unitarily generated operators as expected.

**Benefits:**
 This makes the attribute `has_unitary_generator` more intuitive, reducing the likelihood of misunderstanding. The keyword argument `allow_nonunitary=False` in `qml.transforms.expand_nonunitary_gen` uses this, so it should now behave as expected. The set of operators considered by the `_recursive_decomposition()` method of `qml.ops.op_math.exp.Exp` is now larger.

**Related GitHub Issues:**
#4055

### CODE REVIEW

File: pennylane/ops/op_math/exp.py [hunk 1]
```diff
@@ -277,9 +276,21 @@ def _recursive_decomposition(self, base: Operator, coeff: complex):
...
-            if op_class.num_wires in {base.num_wires, AnyWires} and op_class is not qml.PauliRot:
+        # Store operator classes with generators
+        has_generator_types = []
+        has_generator_types_anywires = []
+        for op_name in qml.ops.qubit.__all__:  # pylint:disable=no-member
+            op_class = getattr(qml.ops.qubit, op_name)  # pylint:disable=no-member
+            if op_class.has_generator and op_class.num_wires == AnyWires:
+                has_generator_types_anywires.append(op_class)
+            elif op_class.has_generator and op_class.num_wires == base.num_wires:
+                has_generator_types.append(op_class)
```
Comment (albi3ro) [inline pennylane/ops/op_math/exp.py:287]:
```suggestion
            if op_class.has_generator:
                if op_class.num_wires == AnyWires:
                    has_generator_types_anywires.append(op_class)
                elif op_class.num_wires == base.num_wires:
                    has_generator_types.append(op_class)
```

Since `has_generator` is on both of these branches, best to extract it.

**Review by albi3ro — COMMENTED** [review]
Looks great!

Just a few minor comments.

To prevent operators with non-unitary generator, we could potentially write a test that makes sure everything actually has a unitary generator.  

In `tests/ops/qubit/test_attributes.py` we could parametrize over every entry and check
```
op = getattr(qml, entry)(1.2, wires=# something)
assert qml.is_unitary(qml.generator(op, format='observable'))
```

I left a couple comments on the PR, but they are just opinions about combining similar pieces of logic.

**Comment by ejthomas** [comment]
Thanks for the review - I will implement these changes as soon as I'm back at my development machine. Completely agree about the test for unitarity in `test_attributes` too - I'll make sure to add that in.

**Comment by trbromley** [comment]
Hi @ejthomas, shall we take another look?

**Comment by ejthomas** [comment]
My latest revision now includes the unit test and code quality updates mentioned above - with the unit test I have made use of the fact that all are unitary specifically up to a factor of two, as calling `is_unitary()` on these generators without accounting for this factor returns `False` and causes the test to fail. I had hoped maybe the `format="prefactor"` argument might normalize these automatically, but this only worked on a subset of the operators (all but the excitations, which give prefactor 1 and leave the observable only unitary up to a factor of 2).

File: pennylane/ops/op_math/exp.py [hunk 2]
```diff
@@ -279,9 +279,22 @@ def _recursive_decomposition(self, base: Operator, coeff: complex):
...
+        has_generator_types_anywires = []
+        for op_name in qml.ops.qubit.__all__:  # pylint:disable=no-member
+            op_class = getattr(qml.ops.qubit, op_name)  # pylint:disable=no-member
+            if op_class.has_generator:
+                if op_class.num_wires == AnyWires:
+                    has_generator_types_anywires.append(op_class)
+                elif op_class.num_wires == base.num_wires:
+                    has_generator_types.append(op_class)
+        # Ensure op_class.num_wires == base.num_wires before op_class.num_wires == AnyWires
+        has_generator_types += has_generator_types_anywires
```
Comment (eddddddy) [inline pennylane/ops/op_math/exp.py:293]: It looks like `has_generator_types_anywires` is not being used past this line. What's the purpose of defining this variable?

File: pennylane/ops/op_math/exp.py [hunk 3]
```diff
@@ -279,9 +279,22 @@ def _recursive_decomposition(self, base: Operator, coeff: complex):
...
+
+        for op_class in has_generator_types:
+            # PauliRot and PCPhase have different positional args
+            if op_class not in {qml.PauliRot, qml.PCPhase}:
```
Comment (eddddddy) [inline pennylane/ops/op_math/exp.py:297]: Question: why do we exclude `PCPhase` here? It certainly has a single unitary generator.

File: tests/ops/op_math/test_exp.py [hunk 4]
```diff
@@ -385,6 +384,11 @@ def test_sparse_matrix_wire_order_error(self):
 class TestDecomposition:
     """Test the decomposition of the `Exp` gate."""
 
+    # Order of `qml.ops.qubit.__all__` is not reliable, so
+    # must sort for consistent order in testing with multiple
+    # workers
+    all_qubit_operators = sorted(qml.ops.qubit.__all__)  # pylint: disable=no-member
```
Comment (eddddddy) [inline tests/ops/op_math/test_exp.py:390]: What exactly is the issue with having this unsorted?

File: tests/ops/op_math/test_exp.py [hunk 5]
```diff
@@ -435,23 +439,33 @@ def test_decomposition_into_pauli_rot(self, base, base_string):
...
 
         phi = 1.23
 
         wires = [0, 1, 2] if op_class.num_wires is AnyWires else list(range(op_class.num_wires))
 
+        # PauliRot and PCPhase each have an extra required arg
         if op_class is qml.PauliRot:
             op = op_class(phi, pauli_word="XYZ", wires=wires)
+        elif op_class is qml.PCPhase:
+            op = op_class(phi, dim=8, wires=wires)
```
Comment (eddddddy) [inline tests/ops/op_math/test_exp.py:462]: Can we test using `dim < 8` so that this operator is non-trivial?

File: tests/ops/op_math/test_exp.py [hunk 6]
```diff
@@ -435,23 +439,33 @@ def test_decomposition_into_pauli_rot(self, base, base_string):
...
         else:
             op = op_class(phi, wires=wires)
 
         exp = qml.evolve(op.generator(), coeff=-phi)
+        if not exp.has_decomposition:
+            pytest.skip("Operator does not have a decomposition")
```
Comment (eddddddy) [inline tests/ops/op_math/test_exp.py:468]: Why is this line necessary? I'd expect the exponential of a generator (times the parameter) to be equal to the operator in every scenario.

File: tests/ops/qubit/test_attributes.py [hunk 7]
```diff
@@ -471,3 +472,25 @@ def test_pcphase(self):
...
         assert np.allclose(mat1, expected_mat)
         assert np.allclose(mat2, expected_mat)
+
+
+# pylint:disable=too-few-public-methods
+class TestHasUnitaryGenerator:
+    """Test that all operations in the ``has_unitary_generator`` attribute
+    actually have unitary generators."""
+
+    unitarily_generated_ops = sorted(list(has_unitary_generator))
```
Comment (eddddddy) [inline tests/ops/qubit/test_attributes.py:482]: Again, why do we need to sort this list?

**Review by eddddddy — COMMENTED** [review]
Great work @ejthomas 🔥 🔥 🔥 

I mostly just have questions about some bits and pieces, but nothing too major. Happy to approve after these are addressed 🙂

File: tests/ops/op_math/test_exp.py (hunk 4 above)
Comment (ejthomas) [reply tests/ops/op_math/test_exp.py:390]: The sorting is related to the use of `pytest-xdist` to distribute tests across multiple CPU workers when testing on the server side. If the different worker processes receive marked parameters in different orders, a failure is raised in the test. This doesn't occur on a local machine unless the user installs `pytest-xdist` and runs the tests with the `-n 2` argument to ensure parallel testing (the server side testing does something similar automatically). By sorting, parametrized tests always run in the same order, avoiding this problem.

File: tests/ops/qubit/test_attributes.py (hunk 7 above)
Comment (ejthomas) [reply tests/ops/qubit/test_attributes.py:482]: Because the `Attribute` class extends `set`, this list would not have a consistent order without sorting - it would fail if parallel test execution takes place, as described above.

File: pennylane/ops/op_math/exp.py (hunk 3 above)
Comment (ejthomas) [reply pennylane/ops/op_math/exp.py:297]: The original decomposition procedure did not include `PCPhase` in the list of operator classes to iterate over here, but the explicit exclusion of `PauliRot` was already present. `PauliRot` requires an extra argument to instantiate, so the next line fails if it is not excluded. As `PCPhase` also has an extra positional argument it should cause the same error on the next line. I have just tried this however, and the tests do not fail if `PCPhase` is allowed; this means that the following line is never reached with `op_class == PCPhase`. I could allow `PCPhase`, test for that case and give different args for it, if we think that case should be handled here (it wasn't handled previously, so I wasn't sure if it was supposed to be possible).

File: pennylane/ops/op_math/exp.py (hunk 2 above)
Comment (ejthomas) [reply pennylane/ops/op_math/exp.py:293]: I defined this list separately and then appended to `has_generator_types` to ensure that the more general operators with `AnyWires` are checked after the more specific ones. In the previous implementation, this was achieved by using an order specified by hand; here, as the list is constructed at runtime, this ensures these operations are checked last. It's just because to construct this list with only one pass of the loop, I have to store both cases in parallel. I could remove this if I loop through just extracting those without `AnyWires`, and then a second time for the `AnyWires` cases, but this would be slower?

File: pennylane/ops/qubit/attributes.py [hunk 8]
```diff
@@ -146,16 +133,11 @@ def __contains__(self, obj):
...
-        "IsingXY",
         "IsingZZ",
-        "SingleExcitation",
         "SingleExcitationMinus",
         "SingleExcitationPlus",
-        "DoubleExcitation",
         "DoubleExcitationMinus",
         "DoubleExcitationPlus",
-        "OrbitalRotation",
-        "FermionicSWAP",
```
[the remark this answers is not in the document]
Comment (ejthomas) [reply pennylane/ops/qubit/attributes.py:158]: Apologies, it does indeed have a unitary generator. There don't appear to be others missing (I'm writing a test for this). The generator of `PCPhase` is exactly unitary, not just up to a factor of 2. I'll update the test so the factor of 2 is not necessary (I thought previously this factor was a property of the mathematics that could be relied on).

File: doc/releases/changelog-dev.md [hunk 9]
```diff
@@ -280,6 +280,9 @@
 
 * A more meaningful error message is raised when broadcasting with adjoint differentation on `DefaultQubit`.
   [(#4203)](https://github.com/PennyLaneAI/pennylane/pull/4203)
+  
+* The `has_unitary_generator` attribute in `qml.ops.qubit.attributes` no longer contains operators with non-unitary generators.
+  [(#4055)](https://github.com/PennyLaneAI/pennylane/pull/4183)
```
[the remark this answers is not in the document]
Comment (ejthomas) [reply doc/releases/changelog-dev.md:285]: Thanks for pointing out (I mixed up issue number with PR number)! Do I need to add my name to the list of contributors to the dev build just below in the changelog, or is that done separately?

File: tests/ops/op_math/test_exp.py (hunk 5 above)
Comment (ejthomas) [reply tests/ops/op_math/test_exp.py:462]: Ah of course - thanks for spotting!

File: tests/ops/op_math/test_exp.py (hunk 6 above)
Comment (ejthomas) [reply tests/ops/op_math/test_exp.py:468]: Without this, if a suitable decomposition is not found by the loop, the test fails - there were some phase operators (`PCPhase`, `CPhase`, `PhaseShift` etc.) which were not previously included in the loop as possible decompositions or tested in the tests that triggered this error. I (probably mistakenly) assumed that they were not intended to be handled by the decomposition function, due to the error and their not being in the tests, so I also skipped them. To handle these, I might need a little guidance on how these operators should decompose here?

**Comment by ejthomas** [comment]
Many thanks for the feedback - I've responded to each point above. I've fixed a number of the issues locally and will update the PR soon. Where the `PCPhase` operator was not handled, I'd benefit from guidance on what the desired behaviour is if anyone has the available bandwidth?

File: doc/releases/changelog-dev.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (eddddddy) [reply doc/releases/changelog-dev.md:285]: Yes, add your name to the list of contributors.

File: pennylane/ops/op_math/exp.py (hunk 2 above)
Comment (eddddddy) [reply pennylane/ops/op_math/exp.py:293]: Ah I see, I missed the comment in the line above. No that should be fine.

File: pennylane/ops/op_math/exp.py (hunk 3 above)
Comment (eddddddy) [reply pennylane/ops/op_math/exp.py:297]: I see. These edge cases we can always think about later. For this PR (and unitaryhack) this is sufficient.

File: pennylane/ops/op_math/exp.py [hunk 10]
```diff
@@ -279,9 +279,22 @@ def _recursive_decomposition(self, base: Operator, coeff: complex):
...
+        # Store operator classes with generators
+        has_generator_types = []
+        has_generator_types_anywires = []
+        for op_name in qml.ops.qubit.__all__:  # pylint:disable=no-member
+            op_class = getattr(qml.ops.qubit, op_name)  # pylint:disable=no-member
+            if op_class.has_generator:
+                if op_class.num_wires == AnyWires:
+                    has_generator_types_anywires.append(op_class)
+                elif op_class.num_wires == base.num_wires:
+                    has_generator_types.append(op_class)
```
Comment (eddddddy) [inline pennylane/ops/op_math/exp.py:291]: I think this should fix most of the issues in the failing `test_generator_decomposition` test.
```suggestion
                elif op_class.num_wires == len(base.wires):
                    has_generator_types.append(op_class)
```

File: tests/ops/op_math/test_exp.py (hunk 6 above)
Comment (eddddddy) [reply tests/ops/op_math/test_exp.py:468]: See my suggestion below. Most of these test failures are for ops whose generator is a `qml.Projector`, which has `num_wires=AnyWires`. Thus `Exp.decomposition` previously skips most of the ops in `qml.ops.qubit`.

File: pennylane/ops/op_math/exp.py (hunk 10 above)
Comment (ejthomas) [reply pennylane/ops/op_math/exp.py:291]: Thanks, that fixes those!

File: tests/ops/op_math/test_exp.py (hunk 6 above)
Comment (ejthomas) [reply tests/ops/op_math/test_exp.py:468]: Thanks - I'm skipping the `PCPhase` test and comparing the generator for `PhaseShift` and `U1` to both of these classes.

**Review by eddddddy — APPROVED** [review]
Great work @ejthomas 💯 

Just a few minor comments but leaving my approval now. Of course this is conditional on these comments being addressed.

File: pennylane/ops/op_math/exp.py [hunk 11]
```diff
@@ -279,9 +279,22 @@ def _recursive_decomposition(self, base: Operator, coeff: complex):
...
+        has_generator_types_anywires = []
+        for op_name in qml.ops.qubit.__all__:  # pylint:disable=no-member
+            op_class = getattr(qml.ops.qubit, op_name)  # pylint:disable=no-member
+            if op_class.has_generator:
+                if op_class.num_wires == AnyWires:
+                    has_generator_types_anywires.append(op_class)
+                elif op_class.num_wires == len(base.wires):
+                    has_generator_types.append(op_class)
+        # Ensure op_class.num_wires == base.num_wires before op_class.num_wires == AnyWires
+        has_generator_types += has_generator_types_anywires
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/ops/op_math/exp.py:293]: I think both work fine.

**Review by albi3ro — APPROVED** [review]
Looks good to me!

Thanks for your contributions :)

File: tests/ops/op_math/test_exp.py [hunk 12]
```diff
@@ -454,7 +469,16 @@ def test_generator_decomposition(self, op_class):
...
         assert len(dec) == 1
-        assert qml.equal(op, dec[0])
+        if op_class in {qml.PhaseShift, qml.U1}:
+            # These operators have the same generator so when reconstructing from
+            # the generator, cannot predict which will be returned
+            assert (
+                isinstance(dec[0], (qml.PhaseShift, qml.U1))
+                and qml.math.isclose(dec[0].data[0], phi)
+                and dec[0].wires == op.wires
+            )
```
[the remark this answers is not in the document]
Comment (ejthomas) [reply tests/ops/op_math/test_exp.py:479]: Thanks!

File: tests/ops/qubit/test_attributes.py [hunk 13]
```diff
@@ -471,3 +472,61 @@ def test_pcphase(self):
...
+            op = op_class(phi, wires=wires)
+        gen = qml.generator(op, format="observable")
+        assert not qml.math.isclose(
+            qml.matrix(qml.prod(qml.adjoint(gen), gen))[0, 0], 0
+        )  # Not unitary if diagonal contains zeros
+        # Some generators are unitary up to a factor - in this case product
+        # of generator with its adjoint has square of this factor on
+        # diagonal
+        coeff = 1.0 / np.sqrt(qml.matrix(qml.prod(qml.adjoint(gen), gen))[0, 0])
+        assert qml.is_unitary(qml.s_prod(coeff, gen))
```
[the remark this answers is not in the document]
Comment (ejthomas) [reply tests/ops/qubit/test_attributes.py:506]: Thanks for your help, this way is a lot clearer! Because `qml.is_unitary` expects an `Operator` rather than an array/matrix, I couldn't see a way to perform the column-wise division, but if I understand correctly, to handle the case of unitarity up to a factor we should only need a single column norm? If the first column in $G$ has norm $|k|$, the first on-diagonal term of the product $G^\dagger G$ is $|k^2|$; if dividing by $|k|$ doesn't give a unitary matrix there can't be another constant which does (as $G^\dagger G \ne \mathbb{I}$). Let me know if there's any issue with this reasoning?

[the remark this answers is not in the document]
Comment (eddddddy) [reply tests/ops/qubit/test_attributes.py:506]: Sure, I guess only the first norm is necessary. Then the coefficient is simply `np.linalg.norm(qml.matrix(gen)[0])`.

File: pennylane/ops/op_math/exp.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (ejthomas) [reply pennylane/ops/op_math/exp.py]: Thanks! I've updated this.

**Comment by ejthomas** [comment]
Many thanks for all your help!

**Review by eddddddy — APPROVED** [review]
Thanks for this @ejthomas 😄 

The CI is kinda on fire right now due to the volume of PRs, but we'll make sure this gets merged before the unitaryhack deadline.
